datafusion/src/sql/planner.rs

```diff
@@ -372,25 +373,91 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
                 // extract join keys
                 extract_join_keys(&expr, &mut keys, &mut filter);
 
+                let mut cols = HashSet::new();
+                exprlist_to_columns(&filter, &mut cols)?;
+
                 let (left_keys, right_keys): (Vec<Column>, Vec<Column>) =
                     keys.into_iter().unzip();
-                // return the logical plan representing the join
-                let join = LogicalPlanBuilder::from(left).join(
-                    right,
-                    join_type,
-                    (left_keys, right_keys),
-                )?;
 
+                // return the logical plan representing the join
                 if filter.is_empty() {
+                    let join = LogicalPlanBuilder::from(left).join(
+                        &right,
+                        join_type,
+                        (left_keys, right_keys),
+                    )?;
                     join.build()
                 } else if join_type == JoinType::Inner {
+                    let join = LogicalPlanBuilder::from(left).join(
+                        &right,
+                        join_type,
+                        (left_keys, right_keys),
+                    )?;
                     join.filter(
                         filter
                             .iter()
                             .skip(1)
                             .fold(filter[0].clone(), |acc, e| acc.and(e.clone())),
                     )?
                     .build()
+                }
+                // Left join with all non-equijoin expressions from the right
```

Comment:
I am curious for left join, is it valid to have the non-equijoin filter expression filter on columns from the left side as well?

Reply:
If you would push the filters to the left logical plan, it would filter out the left keys.

For example:


```
left | right
1    |   2
2    |   4
```

if we would push the filter `left > 2` to the left side it would filter out that row, but the left join requires to result in all left keys (and only match / not match to values on the right).

Reply:
An option to support non-equijoin filter expression for left/left and right/right combinations would be to support filter expression in the join implementation.

For efficiency, for expressions only referring to the left (or right for right join) side, this can be done before loading those values into the hashmap and directly produce values for the not matching left + null values for the right - this saves memory / time if the additional condition filters out a lot.

Filters that include both the left and right columns (e.g.  `left_col > right_col`) could be done inside the probing loop when producing the (left, right) tuples or as some "post processing" by nullifying the non-matching rows on the right side.

Reply:
> I am curious for left join, is it valid to have the non-equijoin filter expression filter on columns from the left side as well?

Yes, as @Dandandan  mentioned it is perfectly valid (and DataFusion today reasonably says "not supported")

However, to implement the semantics correctly, again as @Dandandan , you have to implement the predicate within the join itself. 

The reason for this is if you have a query like `A LEFT JOIN B` the semantics mean *output at least one entry for all rows in A; If the ON condition evaluates to FALSE then output the row from A with the values for B set to NULL"

The technique in this PR is to recognize you can push filtering on the non-preserved side below the join using a filter

Reply:
Thanks @Dandandan and @alamb for the detailed and clear explanation, makes total sense to me :+1: